flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/wmassigners/MiniBatchedWatermarkAssignerOperator.java

```diff
@@ -1,208 +0,0 @@
-			int rowtimeFieldIndex,
-			WatermarkGenerator watermarkGenerator,
-			long idleTimeout,
-			long minibatchInterval) {
-		checkArgument(minibatchInterval > 0, "The inferred emit latency should be larger than 0");
-		this.rowtimeFieldIndex = rowtimeFieldIndex;
-		this.watermarkGenerator = watermarkGenerator;
-		this.idleTimeout = idleTimeout;
-		this.minibatchInterval = minibatchInterval;
-		this.chainingStrategy = ChainingStrategy.ALWAYS;
-	}
-
-	@Override
-	public void open() throws Exception {
-		super.open();
-
-		currentWatermark = 0;
-		nextWatermark = getMiniBatchStart(currentWatermark, minibatchInterval) + minibatchInterval - 1;
-
-		if (idleTimeout > 0) {
-			this.lastRecordTime = getProcessingTimeService().getCurrentProcessingTime();
-			this.streamStatusMaintainer = getContainingTask().getStreamStatusMaintainer();
-			getProcessingTimeService().registerTimer(lastRecordTime + idleTimeout, this);
-		}
-
-		FunctionUtils.setFunctionRuntimeContext(watermarkGenerator, getRuntimeContext());
-		FunctionUtils.openFunction(watermarkGenerator, new Configuration());
-	}
-
-	@Override
-	public void processElement(StreamRecord<BaseRow> element) throws Exception {
-		if (idleTimeout > 0) {
-			// mark the channel active
-			streamStatusMaintainer.toggleStreamStatus(StreamStatus.ACTIVE);
-			lastRecordTime = getProcessingTimeService().getCurrentProcessingTime();
-		}
-		BaseRow row = element.getValue();
-		if (row.isNullAt(rowtimeFieldIndex)) {
-			throw new RuntimeException("RowTime field should not be null," +
-				" please convert it to a non-null long value.");
-		}
-		Long watermark = watermarkGenerator.currentWatermark(row);
-		if (watermark != null) {
-			currentWatermark = Math.max(currentWatermark, watermark);
-		}
-		// forward element
-		output.collect(element);
-
-		// emit watermark if reach to the next watermark
-		if (currentWatermark >= nextWatermark) {
-			advanceWatermark();
-		}
-	}
-
-	private void advanceWatermark() {
-		output.emitWatermark(new Watermark(currentWatermark));
-		long start = getMiniBatchStart(currentWatermark, minibatchInterval);
-		long end = start + minibatchInterval - 1;
-		nextWatermark = end > currentWatermark ? end : end + minibatchInterval;
-	}
-
-	/**
-	 * The processing time trigger is only used for idle timeout. This will not flush watermarks
-	 * because watermark advancing is all controlled by {@link #processElement(StreamRecord)}.
-	 */
-	@Override
-	public void onProcessingTime(long timestamp) throws Exception {
-		if (idleTimeout > 0) {
-			final long currentTime = getProcessingTimeService().getCurrentProcessingTime();
-			if (currentTime - lastRecordTime > idleTimeout) {
-				// mark the channel as idle to ignore watermarks from this channel
-				streamStatusMaintainer.toggleStreamStatus(StreamStatus.IDLE);
-			}
-		}
-
-		// register next timer
-		long now = getProcessingTimeService().getCurrentProcessingTime();
-		getProcessingTimeService().registerTimer(now + idleTimeout, this);
```

Comment:
I've noticed that in `WatermarkAssignerOperator`, timer interval is `registerTimer(now + watermarkInterval, this)`;
I have a concern that if idle timeout < watermarkInterval, we may not toggle the current stream status in time?

Reply:
You are right. This is a limitation for now, because we can't register different timers. We should improve this in the future .